uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -0,0 +1,642 @@
+//! transform the result into a `Result`.
+//!
+//! This type may not be instantiated directly, but _via_ the procedural macros,
+//! such as `#[uniffi::export]`. A `RustFuture` is created, boxed, and then
+//! manipulated by (hidden) helper functions, resp. [`uniffi_rustfuture_poll`]
+//! and [`uniffi_rustfuture_drop`]. Because the `RustFuture` type contains a
+//! generic parameters `T` and `E`, the procedural macros will do a
+//! monomorphisation phase so that all the API has all their types statically
+//! known.
+//!
+//! # The big picture
+//!
+//! This section will explain how the entire workflow works.
+//!
+//! Let's consider the following Rust function:
+//!
+//! ```rust,ignore
+//! #[uniffi::export]
+//! async fn hello() -> bool {
+//!     true
+//! }
+//! ```
+//!
+//! In Rust, this `async fn` syntax is strictly equivalent to:
+//!
+//! ```rust,ignore
+//! #[uniffi::export]
+//! fn hello() -> impl Future<Output = bool> { /* … */ }
+//! ```
+//!
+//! Once this is understood, it becomes obvious that an `async` function
+//! returns a `Future`.
+//!
+//! This function will not be modified, but new functions with a C ABI will be
+//! created, as such:
+//!
+//! ```rust,ignore
+//! /// The `hello` function, as seen from the outside. It returns a `Future`, or
+//! /// more precisely, a `RustFuture` that wraps the returned future.
+//! #[no_mangle]
+//! pub extern "C" fn _uniffi_hello(
+//!     call_status: &mut ::uniffi::RustCallStatus
+//! ) -> Option<Box<::uniffi::RustFuture<bool, ::std::convert::Infallible>>> {
+//!     ::uniffi::call_with_output(call_status, || {
+//!         Some(Box::new(::uniffi::RustFuture::new(async move {
+//!             Ok(hello().await)
+//!         })))
+//!     })
+//! }
+//! ```
+//!
+//! The second generated function is the _poll function_:
+//!
+//! ```rust,ignore
+//! /// The function to poll the `RustFuture` returned by `_uniffi_hello`.
+//! #[no_mangle]
+//! pub extern "C" fn _uniffi_hello_poll(
+//!     future: Option<&mut ::uniffi::RustFuture<bool, ::std::convert::Infallible>>,
+//!     waker: Option<NonNull<::uniffi::RustFutureForeignWakerFunction>>,
+//!     waker_environment: *const ::uniffi::RustFutureForeignWakerEnvironment,
+//!     polled_result: &mut <bool as ::uniffi::FfiReturn>::FfiType,
+//!     call_status:: &mut ::uniffi::RustCallStatus,
+//! ) -> bool {
+//!     ::uniffi::ffi::uniffi_rustfuture_poll(future, waker, waker_environment, polled_result, call_status)
+//! }
+//! ```
+//!
+//! Let's analyse this function because it's an important one:
+//!
+//! * First off, this _poll FFI function_ forwards everything to
+//!   [`uniffi_rustfuture_poll`]. The latter is generic, while the former has been
+//!   monomorphised by the procedural macro.
+//!
+//! * Second, it receives the `RustFuture` as an `Option<&mut RustFuture<_>>`. It
+//!   doesn't take ownership of the `RustFuture`! It borrows it (mutably). It's
+//!   wrapped inside an `Option` to check whether it's a null pointer or not.
```

Comment:
- Maybe mention that the future comes from `_uniffi_hello` here
- Also mention why `null` would be sent over the FFI.  Is this just defensive programming or is there a use-case for the foreign side sending a `null` value?

Reply:
Done.

`null` is indeed accepted for defensive programming purposes, so that Rust can panic gracefully, with `CallStatus`, exceptions and errors on the foreign side etc.